### PR TITLE
was/templates

### PR DESCRIPTION
**Pull request by parvdesai-tenable**
was/templates broken into two parts here
1. templates
2. user-templates
Contains changes related to those endpoints

**The WAS Search method and iterator depends upon #69**

### CODE REVIEW

File: tenable/io/v3/was/__init__.py [hunk 1]
```diff
@@ -17,11 +17,15 @@
     :glob:
 
     folders
+    templates
+    user-templates
     scans
```
Comment (milanaghera-tenable) [inline tenable/io/v3/was/__init__.py:22]: Please follow the ASCII order. So, Add agent_groups after agent_config, similar to the order in imports.

File: tenable/io/v3/was/__init__.py [hunk 2]
```diff
@@ -45,3 +49,19 @@ def scans(self):
         :doc:`Scans API <scans>`
         '''
         return ScansAPI(self._api)
+
+    @property
+    def templates(self):
+        '''
+        The interface object for the
+        :doc:`Folders API <folders>`
```
Comment (milanaghera-tenable) [inline tenable/io/v3/was/__init__.py:57]: For consistency, Please update `Folders API <folders>` to `Tenable.io Folders API <folders>`

File: tenable/io/v3/was/__init__.py [hunk 3]
```diff
@@ -45,3 +49,19 @@ def scans(self):
...
+        '''
+        return TemplatesAPI(self._api)
+
+    @property
+    def user_templates(self):
+        '''
+        The interface object for the
+        :doc:`Folders API <folders>`
```
Comment (milanaghera-tenable) [inline tenable/io/v3/was/__init__.py:65]: For consistency, Please update Folders API <folders> to Tenable.io Folders API <folders>

File: tenable/io/v3/was/user_templates/schema.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+from marshmallow import Schema, fields, validate
```
Comment (milanaghera-tenable) [inline tenable/io/v3/was/user_templates/schema.py:1]: Please Add the docstring here.
for example,
'''
User Template API Endpoint Schemas
'''

File: tenable/io/v3/was/user_templates/schema.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+
+class PermissionSchema(Schema):
+    entity = fields.Str(
```
Comment (milanaghera-tenable) [inline tenable/io/v3/was/user_templates/schema.py:5]: Please add docstring here

File: tenable/io/v3/was/user_templates/schema.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    entity_id = fields.UUID(required=True)
+    level = fields.Str(
+        required=True,
+        validate=validate.OneOf(['no_access', 'view', 'configure', 'control'])
+    )
+    permissions_id = fields.UUID(required=True)
+
+
+class UserTemplateSchema(Schema):
+    name = fields.Str()
```
Comment (milanaghera-tenable) [inline tenable/io/v3/was/user_templates/schema.py:18]: Please add docstring here

File: tests/io/v3/was/templates/test_templates_api.py [hunk 7]
```diff
@@ -0,0 +1,452 @@
+import pytest
```
Comment (milanaghera-tenable) [inline tests/io/v3/was/templates/test_templates_api.py:1]: Please add docstring here

File: tests/io/v3/was/user_templates/test_user_templates_api.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
+import pytest
```
Comment (milanaghera-tenable) [inline tests/io/v3/was/user_templates/test_user_templates_api.py:1]: Please add docstring here

File: tests/io/v3/was/user_templates/test_user_templates_schema.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
+from tenable.io.v3.was.user_templates.schema import (PermissionSchema,
```
Comment (milanaghera-tenable) [inline tests/io/v3/was/user_templates/test_user_templates_schema.py:1]: Please add docstring here

File: tests/io/v3/was/user_templates/test_user_templates_schema.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    'default_permissions': 'no_access',
+    'permissions': [permission_obj],
+    'description': 'edited description',
+    'owner_id': 'a8ff6f60-a7e0-43a5-a9d4-0bd079e1d9fa',
+    'results_visibility': 'private'
+}
+
+
+def test_user_template_schema():
+    schema = UserTemplateSchema()
```
Comment (milanaghera-tenable) [inline tests/io/v3/was/user_templates/test_user_templates_schema.py:22]: Please add docstring here

File: tests/io/v3/was/user_templates/test_user_templates_schema.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    assert user_template_obj == schema.dump(schema.load(user_template_obj))
+
+
+def test_permission_schema():
+    schema = PermissionSchema()
```
Comment (milanaghera-tenable) [inline tests/io/v3/was/user_templates/test_user_templates_schema.py:27]: Please add docstring here

**Review by milanaghera-tenable — APPROVED** [review]

File: tenable/io/v3/was/user_templates/api.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+        '''
+        Updates the specified user-defined template.
+
+        :devportal:`was templates: Update user-defined template
+        <was-v2-user-templates-update>`
+
+        Args:
+            user_template_id (UUID): The UUID of the user-defined template.
+            name (str, optional): The name of the user-defined template.
+            owner_id (UUID, optiona;):
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/was/user_templates/api.py:95]: typo: optional

File: tenable/io/v3/was/user_templates/schema.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
+'''
+User Templates API Endpoint Schemas
+'''
+from marshmallow import Schema, fields, validate
+
+
+class PermissionSchema(Schema):
+    '''
+    Permission Schema for user-tempaltes
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/was/user_templates/schema.py:9]: Typo
Permission Schema for User-Templates

**Review by dhvanilshah-tenable — APPROVED** [review]

File: tenable/io/v3/was/user_templates/api.py (hunk 12 above)
Comment (parvdesai-tenable) [reply tenable/io/v3/was/user_templates/api.py]: fixed

File: tenable/io/v3/was/user_templates/schema.py (hunk 13 above)
Comment (parvdesai-tenable) [reply tenable/io/v3/was/user_templates/schema.py]: fixed

**Review by alpesh-te — APPROVED** [review]
LGTM

**Pull request merged**